keps/sig-api-machinery/5589-gogo-dependency/README.md

```diff
@@ -0,0 +1,638 @@
+repos, and scans of public github did not uncover use of these with Kubernetes API types specifically.
+
+**Risk: marshal / unmarshal code is unintentionally modified**
+
+Mitigations:
+* This KEP prioritizes leaving the in-use marshal / unmarshal code paths untouched,
+  focusing on straight removal of unused code paths and imports instead,
+  or modification of generators that leave generated code unmodified.
+* Existing compatibility tests ensure protobuf data from previous releases is round-tripped losslessly
+
+## Design Details
+
+**Truncate exported methods**
+
+Only the exported methods used by `k8s.io/apimachinery` protobuf handling or with other widespread use will be kept:
+
+* Unmarshaling:
+    ```go
+    Reset()
+    Unmarshal([]byte) error
+    ```
+* Marshaling:
+    ```go
+    Size() int
+    Marshal() ([]byte, error)
+    MarshalTo(data []byte) (int, error)
+    MarshalToSizedBuffer(data []byte) (int, error)
+    ```
+
+* Other
+    ```go
+    String() string
+    ```
+
+We will truncate the exported methods from protobuf generation to just those methods.
+
+**Truncate imports**
+
+The only gogo packages used in generated code for anything other than a type assertion are:
+
+* The `sortkeys` package, which is a pass-through to the stdlib `sort.Strings`.
+  This will be replaced with a direct call to stdlib `sort.Strings`.
+* The `proto` package, used for `init()`-time registration of types and file descriptors.
+  The `init()` time global registrations are not required by apimachinery and will be dropped.
+
+**Truncate exported variables**
+
+Only the exported variables used by retained methods will be kept:
+* `ErrInvalidLengthGenerated`
+* `ErrIntOverflowGenerated`
+* `ErrUnexpectedEndOfGroupGenerated`
+
+**Truncate unused private variables and methods**
+
+After the truncation of exported methods, variables, and `init()`-time registration,
+other imports, variables, and methods which are now unused will be dropped.
+
+**Fork gogo generator**
+
+The subset of the gogo code generation used by `k8s.io/code-generator/cmd/go-to-protobuf`
+will be forked to a location within the `kubernetes` or `kubernetes-sigs` project,
+and pruned / modified to only output the subset of generated code kept by `go-to-protobuf`.
+
+**Modify String() implementation**
+
+There is widespread reliance on Kubernetes API objects implementing `String()`,
+so preserving the existence of `String()` implementations is necessary to minimize disruption.
+
+However, this only *happens* to be done as part of protobuf generation,
+and the output is a pseudo-Go structure that requires large amounts of code to construct,
+and is not particularly meaningful from an API perspective.
+
+After the code-generation is forked and modifiable, the `String()` implementation will be
```

Comment:
is this a separate and independent step?
should this be feature gated just in case?

Reply:
my concern is that this cosmetic thing just found a project foo that depends on this output and we have to make a choice on reverting the whole thing, this looks more like a nice to have that adds a 1% of the value of this entire kep

Reply:
yeah... if that's a distraction I can remove the plan to modify this and lump that into possible changes we might do in the future separate from executing this KEP

Reply:
moved the `String()` simplification to a "future work" section